services/src/main/java/org/keycloak/authentication/actiontoken/idpverifyemail/IdpVerifyAccountLinkActionTokenHandler.java

```diff
@@ -133,6 +139,37 @@ public Response handleToken(IdpVerifyAccountLinkActionToken token, ActionTokenCo
         return tokenContext.brokerFlow(null, null, authSession.getAuthNote(AuthenticationProcessor.CURRENT_FLOW_PATH));
     }
 
+    private void setEmailVerifiedSingleObject(IdpVerifyAccountLinkActionToken token, RealmModel realm, KeycloakSession session, UserModel user) {
+        int singleObjectLifespan = realm.getActionTokenGeneratedByUserLifespan();
+        String userId = user.getId();
+        String idpAlias = token.getIdentityProviderAlias();
+        session.singleUseObjects().put(getEmailVerifiedSingleObjectKey(userId, idpAlias), singleObjectLifespan, Map.of());
+    }
+
+    public static boolean isEmailVerified(KeycloakSession session, UserModel user, IdentityProviderModel broker, Runnable runnable) {
+        if (user == null) {
+            return false;
+        }
+
+        SingleUseObjectProvider singleObjects = session.singleUseObjects();
+        String singleObjectKey = getEmailVerifiedSingleObjectKey(user.getId(), broker.getAlias());
+        boolean isEmailVerified = singleObjects.get(singleObjectKey) != null;
+
+        if (isEmailVerified) {
+            try {
+                runnable.run();
+            } finally {
+                singleObjects.remove(singleObjectKey);
+            }
+        }
+
+        return isEmailVerified;
+    }
+
+    private static String getEmailVerifiedSingleObjectKey(String userId, String idpAlias) {
+        return "emailVerified" + userId + idpAlias;
```

Comment:
```suggestion
        return "verified" + userId + idpAlias;
```

Maybe if it is removed to `verified`, it can be better as the "signal" that user verified his account may not be specific to email, but could be used by other use-cases in theory. Similarly the methods could be renamed to `isUserVerified` instead of `isEmailVerified` .

But this is possibly just a detail considering that we do not have more OOTB use-cases besides email verification right now.